# Converting CSV Files to LAS Files Using LASIO
*Created by: Andy McDonald*  
*Article Version:*  
*YouTube Video:*  

Well log data can be delivered in a variety of formats (DLIS, LAS, CSV, ASC etc.). however, if you have a CSV file containing well log data, and you want to convert it to LAS format there are a number of ways you can achieve this. So if you want to see how to do it using the LASIO library then keep watching.

One of the issues you may come across when looking at well log data is the large variety of data formats. Data can be delivered or stored in DLIS files, LAS Files, CSV Files and many other different formats. This can often become a headache trying to work out which file to use or when you come to creating a standard file that can be easily read by commercial software or by other users.

It is a simple format and it is a flat file that contains meta data about the well within the header section, and parameter information, as well as the log data measurements. These files are easy to open up in any text editor and you can quickly and easily read the content. 

However there may be occasions where you end up with a CSV file containing well log measurements and you want to convert it to a LAS file. Well, that is what we are going cover in todays video. 

We are going to see how we can take a simple CSV file like this and convert it to a LAS file like this using the excellent LASIO library.

## Load Required Libraries
First off, we will load in the required python libraries. For this tutorial we will be using lasio and pandas.

For more information about using lasio, check out this article here: https://andymcdonaldgeo.medium.com/loading-and-displaying-well-log-data-b9568efd1d8

In [1]:
import lasio
import pandas as pd

## Load in CSV File Using Pandas

Next we need to load in our csv file.
This is done using the `read_csv()` function from pandas and passing in the file location and file name.

In [2]:
data = pd.read_csv('Data/Notebook 22/Notebook 22 - VOLVE - 15_9-19.csv')

Once the file has been read, we can then check what the contents of the file are  by calling upon the pandas `.head()` function.

In [3]:
data.head()

DEPTH   CALI  COAL       DT   DT_LOG       DTS   DTS_LOG      GR  \
0  3500.0183  9.315     0  76.7292  76.7292  157.1754  157.1754  36.621   
1  3500.1707  9.324     0  77.2473  77.2473  158.9566  158.9566  36.374   
2  3500.3231  9.338     0  77.8462  77.8462  159.7642  159.7642  30.748   
3  3500.4755  9.329     0  78.3571  78.3571  158.7547  158.7547  29.795   
4  3500.6279  9.328     0  78.6560  78.6560  157.1320  157.1320  27.346   

     NPHI    PHIE   PHIEC    PHIT   PHITC    RHOB  RHOB_LOG     RT      RW  \
0  0.1542  0.1122  0.1098  0.1209  0.1186  2.4602     2.460  1.791  0.0211   
1  0.1694  0.1074  0.1060  0.1159  0.1146  2.4680     2.468  1.756  0.0211   
2  0.1776  0.1082  0.1079  0.1127  0.1125  2.4730     2.473  1.720  0.0211   
3  0.1767  0.1254  0.1226  0.1292  0.1264  2.4471     2.447  1.696  0.0211   
4  0.1662  0.1278  0.1249  0.1299  0.1270  2.4460     2.446  1.697  0.0211   

      TEMP  
0  94.5855  
1  94.5897  
2  94.5940  
3  94.5982  
4  94.6025

We can now see that we have 18 columns within our dataframe, and a mixture of well log measurements.

To ensure that the data is all numeric and to understand how many nulls are present within the data we can call upon the pandas function `.info()`. This is not a necessary step, but it does allow us to check that the columns are numeric (either float64 or int64).

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4101 entries, 0 to 4100
Data columns (total 18 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   DEPTH     4101 non-null   float64
 1   CALI      4101 non-null   float64
 2   COAL      4101 non-null   int64  
 3   DT        4101 non-null   float64
 4   DT_LOG    4101 non-null   float64
 5   DTS       4101 non-null   float64
 6   DTS_LOG   4101 non-null   float64
 7   GR        4068 non-null   float64
 8   NPHI      4100 non-null   float64
 9   PHIE      4101 non-null   float64
 10  PHIEC     4101 non-null   float64
 11  PHIT      4101 non-null   float64
 12  PHITC     4101 non-null   float64
 13  RHOB      4101 non-null   float64
 14  RHOB_LOG  4101 non-null   float64
 15  RT        4101 non-null   float64
 16  RW        4101 non-null   float64
 17  TEMP      4101 non-null   float64
dtypes: float64(17), int64(1)
memory usage: 576.8 KB


## Create an Empty LAS Object

Before we can transfer our data from CSV to LAS, we first need to create a blank LAS file. This is achieved by calling upon `lasio.LASFile()` and assigning it to a variable. In this example the variable is called `las_file`.

In [5]:
las_file = lasio.LASFile()

When we try to view the contents of the newly created LAS file we can see that the header information is empty.

In [6]:
las_file.header

{'Version': [HeaderItem(mnemonic="VERS", unit="", value="2.0", descr="CWLS log ASCII Stan...),
  HeaderItem(mnemonic="WRAP", unit="", value="NO", descr="One line per depth s...),
  HeaderItem(mnemonic="DLM", unit="", value="SPACE", descr="Column Data Sectio...)],
 'Well': [HeaderItem(mnemonic="STRT", unit="m", value="nan", descr="START DEPTH"),
  HeaderItem(mnemonic="STOP", unit="m", value="nan", descr="STOP DEPTH"),
  HeaderItem(mnemonic="STEP", unit="m", value="nan", descr="STEP"),
  HeaderItem(mnemonic="NULL", unit="", value="-9999.25", descr="NULL VALUE"),
  HeaderItem(mnemonic="COMP", unit="", value="", descr="COMPANY"),
  HeaderItem(mnemonic="WELL", unit="", value="", descr="WELL"),
  HeaderItem(mnemonic="FLD", unit="", value="", descr="FIELD"),
  HeaderItem(mnemonic="LOC", unit="", value="", descr="LOCATION"),
  HeaderItem(mnemonic="PROV", unit="", value="", descr="PROVINCE"),
  HeaderItem(mnemonic="CNTY", unit="", value="", descr="COUNTY"),
  HeaderItem(mnemonic="STAT", unit=""

We can also confirm that we have no data within the file by calling upon `las_file.curves`.

In [7]:
las_file.curves

[]

## Setting up the Metadata
Now that we have a blank las object to work with, we now need to add information to our LAS header. 

The first step is to create a number of variables that we want to fill in. Doing it this way, rather than passing them into the HeaderItem functions makes it easier to change them in the future and also makes it more readable. For instance, if we created a function where we wanted to update specifc parameters within the header based on different files, we can easily pass these variables into the function and we won't have to update the code within the function.

In [8]:
well_name = 'Random Well'
field_name = 'Random Field'
uwi = '123456789'
country = 'Random Country'


To begin assigning the values to the header, we need to call upon `las_file.well` and select the header attribute we want to add to. On the right hand side, we will update the HeaderItem and supply a value to it.

In [9]:
las_file.well['WELL'] = lasio.HeaderItem('WELL', value=well_name)
las_file.well['FLD'] = lasio.HeaderItem('FLD', value=field_name)
las_file.well['UWI'] = lasio.HeaderItem('UWI', value=uwi)
las_file.well['CTRY'] = lasio.HeaderItem('CTRY', value=country)

Once we have done this we can call upon our header again and we can now see that the values for the well name, UWI, country and field name have all been updated.

In [10]:
las_file.header

{'Version': [HeaderItem(mnemonic="VERS", unit="", value="2.0", descr="CWLS log ASCII Stan...),
  HeaderItem(mnemonic="WRAP", unit="", value="NO", descr="One line per depth s...),
  HeaderItem(mnemonic="DLM", unit="", value="SPACE", descr="Column Data Sectio...)],
 'Well': [HeaderItem(mnemonic="STRT", unit="m", value="nan", descr="START DEPTH"),
  HeaderItem(mnemonic="STOP", unit="m", value="nan", descr="STOP DEPTH"),
  HeaderItem(mnemonic="STEP", unit="m", value="nan", descr="STEP"),
  HeaderItem(mnemonic="NULL", unit="", value="-9999.25", descr="NULL VALUE"),
  HeaderItem(mnemonic="COMP", unit="", value="", descr="COMPANY"),
  HeaderItem(mnemonic="WELL", unit="", value="Random Well", descr=""),
  HeaderItem(mnemonic="FLD", unit="", value="Random Field", descr=""),
  HeaderItem(mnemonic="LOC", unit="", value="", descr="LOCATION"),
  HeaderItem(mnemonic="PROV", unit="", value="", descr="PROVINCE"),
  HeaderItem(mnemonic="CNTY", unit="", value="", descr="COUNTY"),
  HeaderItem(mnemonic="

## Adding Curves
To add curves to the file we can use the `add_curve` function and pass in the data and units.

This example here shows how we can add a single curve to the file called DEPT. Note that if adding the main depth data, it does need to go in as DEPT rather than DEPTH.

In [11]:
las_file.add_curve('DEPT', data['DEPTH'], unit='m')

To write out our file, we can call upon `.write()` and pass in the location and file name for our LAS file.

In [13]:
las_file.write('Output/Notebook 22/OutputLAS.las')

## Write Remaining Curves

We can reuse the file that we have created above and write the remaining curves.

To make things easier, I have created a list containing the measurement units for each well log curve. Note that this does include the depth measurement. 

In [14]:
units = ['m',
 'inches',
 'unitless',
 'us/ft',
 'us/ft',
 'us/ft',
 'us/ft',
 'API',
 'v/v_decimal',
 'v/v_decimal',
 'v/v_decimal',
 'v/v_decimal',
 'v/v_decimal',
 'g/cm3',
 'g/cm3',
 'ohm.m',
 'ohm.m',
 'degC']

We can then begin to loop through each of the logging measurements / columns within the dataframe along with the units list. This is achieved using the Python `zip` function.

As we already have depth within our las file, we can skip this column by checking the column name. There are other ways in which this can be handled, for example by writing the depth and curves to the las file in one go.

In [15]:
for col, unit in zip(data.columns, units):
    if col != 'DEPTH':
        las_file.add_curve(col, data[col], unit=unit)
        

When we check the `curves` function, we can see that we have all of our curves and they all have the appropriate units. We can also see from the data.shape part of the listing we have 4101 values per curve which confirms we have data.

In [16]:
las_file.curves

[CurveItem(mnemonic="DEPT", unit="m", value="", descr="", original_mnemonic="DEPT", data.shape=(4101,)),
 CurveItem(mnemonic="CALI", unit="unitless", value="", descr="", original_mnemonic="CALI", data.shape=(4101,)),
 CurveItem(mnemonic="COAL", unit="us/ft", value="", descr="", original_mnemonic="COAL", data.shape=(4101,)),
 CurveItem(mnemonic="DT", unit="us/ft", value="", descr="", original_mnemonic="DT", data.shape=(4101,)),
 CurveItem(mnemonic="DT_LOG", unit="us/ft", value="", descr="", original_mnemonic="DT_LOG", data.shape=(4101,)),
 CurveItem(mnemonic="DTS", unit="us/ft", value="", descr="", original_mnemonic="DTS", data.shape=(4101,)),
 CurveItem(mnemonic="DTS_LOG", unit="API", value="", descr="", original_mnemonic="DTS_LOG", data.shape=(4101,)),
 CurveItem(mnemonic="GR", unit="v/v_decimal", value="", descr="", original_mnemonic="GR", data.shape=(4101,)),
 CurveItem(mnemonic="NPHI", unit="v/v_decimal", value="", descr="", original_mnemonic="NPHI", data.shape=(4101,)),
 CurveItem

We can confirm that we have values by calling upon of the curves. In the example below, I have called upon GR, and we get an array returned containing the Gamma Ray values, which match the values in the dataframe presented earlier.

In [19]:
las_file['GR']

array([  36.621,   36.374,   30.748, ..., -999.   , -999.   , -999.   ])

Once we are happy with the las file, we can now export it to a file and use it in any other software package.

In [17]:
las_file.write('Output/Notebook 22/OutputLAS_FINAL.las')

# Summary
In this article we have covered how to convert a simple CSV file containing well log / petrophysical measurements into a industry standard LAS (Log ASCII Standard) file. Once you have created a blank LASFile object in lasio, you will be able to manually update the header items with the correct metadata and also update the curves with the correct values.

***Thanks for reading!***

*If you have found this article useful, please feel free to check out my other articles looking at various aspects of Python and well log data. You can also find my code used in this article and others at [GitHub](https://github.com/andymcdgeo).*

*If you want to get in touch you can find me on [LinkedIn](https://www.linkedin.com/in/andymcdonaldgeo/) or at my [website](http://andymcdonald.scot/).*

*Interested in learning more about python and well log data or petrophysics? Follow me on [Medium](https://medium.com/@andymcdonaldgeo).*

If you have enjoyed this article or any others and want to show your appreciate you are welcome to [Buy Me a Coffee](https://www.buymeacoffee.com/andymcdonaldgeo)